### PR TITLE
[merged] Atomic/push.py: Don't prompt for password if token found

### PR DESCRIPTION
**Pull request by baude**
If the user has a token for the registry, we assume the token
is good and do not prompt for a username or password.

### CODE REVIEW

File: Atomic/atomic.py [hunk 1]
```diff
@@ -678,6 +679,24 @@ def get_vulnerable_ids(self):
...
 
+    def get_local_tokens(self):
+        if len(self.local_tokens) < 1:
+            self.local_tokens = self.load_local_tokens()
+        return self.local_tokens
+
+    @staticmethod
+    def load_local_tokens():
+        tokens = {}
+        token_file_name = os.path.expanduser('~/.docker/config.json')
```
Comment (rhatdan) [inline Atomic/atomic.py]: @runcom Does Skopeo just do this, or is it a little more cleaver?  Docker?  Does it look at SUDO_USER?

File: Atomic/atomic.py [hunk 2]
```diff
@@ -678,6 +679,24 @@ def get_vulnerable_ids(self):
...
+        if not os.path.exists(token_file_name):
+            return {}
+        with open(token_file_name) as token_file:
+            token_data = json.load(token_file)
+        for registry in token_data['auths']:
+            tokens[registry] = token_data['auths'][registry]['auth']
```
Comment (rhatdan) [inline Atomic/atomic.py]: Should this be in a try block?  Potential errors in the auth.  If the auths and auth do not exists should we return {}

File: Atomic/atomic.py [hunk 3]
```diff
@@ -678,6 +679,24 @@ def get_vulnerable_ids(self):
...
+        return tokens
+
+
 class AtomicError(Exception):
```
Comment (rhatdan) [inline Atomic/atomic.py]: Remove extra space

File: Atomic/push.py [hunk 4]
```diff
@@ -122,24 +129,29 @@ def push(self):
...
+
         if self.args.pulp:
             return pulp.push_image_to_pulp(self.image, self.args.url,
                                            self.args.username,
                                            self.args.password,
                                            self.args.verify_ssl,
                                            self.d)
 
-        if self.args.satellite:
+        elif self.args.satellite:
```
Comment (rhatdan) [inline Atomic/push.py]: Why else? Previous if returned.

File: Atomic/push.py [hunk 5]
```diff
@@ -122,24 +129,29 @@ def push(self):
...
+        sign = True if self.args.sign_by else False
+
+        registry, _, _ = util.decompose(self.image)
+        if sign and not self.args.satellite and not self.args.pulp:
+            # Check if any local tokens exist
+            if registry not in [x for x in self.load_local_tokens()]:
+                # If we find a token for the registry, we don't
+                # prompt for a username or password
+                prompt()
+
```
Comment (rhatdan) [inline Atomic/push.py]: Why not have this block to after the satellite and pulp checks are run.

File: Atomic/atomic.py (hunk 2 above)
Comment (baude) [inline Atomic/atomic.py]: Added try block for a KeyError exception

File: Atomic/atomic.py (hunk 3 above)
Comment (rhatdan) [inline Atomic/atomic.py]: ok

File: Atomic/push.py [hunk 6]
```diff
@@ -156,6 +160,13 @@ def push(self):
 
         else:
             reg, _, tag = util.decompose(self.image)
+            if sign and not self.args.satellite and not self.args.pulp:
```
Comment (rhatdan) [inline Atomic/push.py]: No need for satellite or pulp check.
No need for else: above either.

File: Atomic/push.py [hunk 7]
```diff
@@ -156,6 +160,13 @@ def push(self):
...
+                # Check if any local tokens exist
+                if reg not in [x for x in self.load_local_tokens()]:
```
Comment (rhatdan) [inline Atomic/push.py]: Shouldn't this be get_local_tokens?

File: Atomic/push.py (hunk 6 above)
Comment (baude) [inline Atomic/push.py]: Geez, sloppy on my part.  Fixed.

File: Atomic/push.py (hunk 7 above)
Comment (baude) [inline Atomic/push.py]: yup!  fixed.

**Pull request closed without merging**